hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithBoundJoin.java

```diff
@@ -0,0 +1,308 @@
+	 *          parameter, use {@link #DEFAULT_BATCH_SIZE}
+	 *
+	 * @param collectExceptions - <code>true</code> if this operator has to
+	 *          collect exceptions (which is handled entirely by one of the
+	 *          super classes); <code>false</code> if the operator should
+	 *          immediately throw every {@link ExecOpExecutionException}
+	 */
+	public ExecOpBindJoinSPARQLwithBoundJoin( final SPARQLGraphPattern query,
+	                                          final SPARQLEndpoint fm,
+	                                          final ExpectedVariables inputVars,
+	                                          final boolean useOuterJoinSemantics,
+	                                          final int batchSize,
+	                                          final boolean collectExceptions ) {
+		super(query, fm, inputVars, useOuterJoinSemantics, batchSize, collectExceptions);
+		pattern = QueryPatternUtils.convertToJenaElement(query);
+	}
+
+	@Override
+	protected NullaryExecutableOp createExecutableReqOp( final Set<Binding> solMaps )
+			throws ExecOpExecutionException
+	{
+		final Element elmt = createUnion(solMaps);
+		final SPARQLGraphPattern pattern = new GenericSPARQLGraphPatternImpl1(elmt);
+		final SPARQLRequest request = new SPARQLRequestImpl(pattern);
+		return new ExecOpRequestSPARQL(request, fm, false);
+	}
+
+	protected Element createUnion( final Iterable<Binding> solMaps )
+			throws ExecOpExecutionException
+	{
+		// Populate the ordered list of solution mappings (used for restoring renamed
+		// vars and restoring the join partner vars)
+		solMaps.forEach(solMapsList::add);
+		
+		// Union element
+		final ElementUnion union = new ElementUnion();
+
+		// Collect the variables bound by the given solution mappings
+		// (which are guaranteed to include all join variables).
+		final Set<Var> joinVars = new HashSet<>();
+		for ( final Binding b : solMaps ) {
+			final Iterator<Var> it = b.vars();
+			while ( it.hasNext() ) {
+				joinVars.add( it.next() );
+			}
+		}
+
+		// Get first non-join variable for renaming
+		for( final Var v : varsInQuery ){
+			if( ! joinVars.contains(v) && query.getCertainVariables().contains(v)){
+				renamedVar = v; // first certain non-join
+				break;
+			}
+		}
+
+		if ( renamedVar == null ) {
+			throw new ExecOpExecutionException("No non-join variable found for renaming", this);
+		}
+
+		// Generate the UNION pattern by iterating over all input solution mappings
+		// and replacing the renamed variable
+		int i = 0;
+		for ( final Binding solMap : solMaps ) {
+			// Create a new pattern in which we replace the variables
+			// with values based on the incoming solution mapping
+			SPARQLGraphPattern patternWithBindings;
+			try {
+				patternWithBindings = query.applySolMapToGraphPattern( new SolutionMappingImpl(solMap) );
+			} catch ( VariableByBlankNodeSubstitutionException e ) {
+				throw new ExecOpExecutionException(e, this);
```

Comment:
Throw an `IllegalArgumentException` instead. The base class should take care that none of the solution mappings passed to `createExecutableReqOp` has a bnode for any of the join variables. Therefore, at this point here, the `VariableByBlankNodeSubstitutionException` should never occur. If it does, there is a bug somewhere.